File: src/2352. Equal Row and Column Pairs.js
/**
 * @param {number[][]} grid
 * @return {number}
 */

export function equalPairs(grid) {
  const n = grid.length,
    trie = {}
  for (let i = 0; i < n; i++) {
    let dict = trie
    for (let j = 0; j < n; j++) {
      const char = grid[i][j]
      if (!dict[char]) dict[char] = {}
      dict = dict[char]
    }
    dict.count = (dict.count || 0) + 1
  }

  let count = 0
  for (let i = 0; i < n; i++) {
    let dict = trie
    for (let j = 0; j < n; j++) {
      const char = grid[j][i]
      if (!dict[char]) break
      dict = dict[char]
    }
    if (dict.count) count += dict.count
  }
  return count
}

File: src/2352. Equal Row and Column Pairs.cpp
class TrieNode {
public:
    int count = 0;
    unordered_map<int, TrieNode*> children;
};

class Solution {
public:
    int equalPairs(vector<vector<int>>& grid) {
        const int n = grid.size();

        auto trie = new TrieNode();
        for (int i = 0; i < n; ++i) {
            auto node = trie;
            for (int j = 0; j < n; ++j) {
                const int value = grid[i][j];
                if (node->children.find(value) == node->children.end()) {
                    node->children[value] = new TrieNode();
                }
                node = node->children[value];
            }
            node->count += 1;
        }

        int ans = 0;
        for (int i = 0; i < n; ++i) {
            auto node = trie;
            for (int j = 0; j < n; ++j) {
                const int value = grid[j][i];
                if (node->children.find(value) == node->children.end()) {
                    break;
                }
                node = node->children[value];
            }
            ans += node->count;
        }
        return ans;
    }
};